tests/unit/IonParserBinaryRawTest.js

```diff
@@ -182,6 +182,120 @@ define([
     readSignedInt([0x7F, 0xFF, 0xFF, 0xFF, 0xFF], maxValueForBits(39));
     readSignedInt([0xFF, 0xFF, 0xFF, 0xFF, 0xFF], -1 * maxValueForBits(39));
 
+    /**
+     * VarUInt
+     *
+     * Spec: http://amzn.github.io/ion-docs/docs/binary.html#varuint-and-varint-fields
+     */
+
+    // The base test function called by the more specific flavors below.
+    // Creates a test which will attempt to read the `expected` VarUInt from the provided input bytes, handling
+    // any anticipated exceptions.
+    let readVarUnsignedIntTest = function(testName, bytes, expected, throwsException) {
+      let test = function() {
+        let binarySpan = new ion.BinarySpan(new Uint8Array(bytes));
+        let actual = ion.ParserBinaryRaw.readVarUnsignedIntFrom(binarySpan);
+        assert.equal(actual, expected)
+      }
+      registerTest(testName, test, throwsException);
+    }
+
+    // Should read the `expected` VarUInt value from the provided input bytes.
+    // Will fail if an exception is thrown or if the value that's read from the bytes is not equal to `expected`.
+    let readVarUnsignedInt = function(bytes, expected) {
+      let testName = 'Read var unsigned int ' + expected + ' from bytes: ' + bytes;
+      let testThrows = false;
+      readVarUnsignedIntTest(testName, bytes, expected, testThrows);
+    }
+
+    // Should detect that overflow has occurred while reading and throw an exception.
+    // Will fail if no exception is thrown.
+    let overflowWhileReadingVarUnsignedInt = function(bytes, expected) {
+      let testName = 'Overflow while attempting to read var unsigned int ' + expected + ' from the bytes: ' + bytes;
+      let testThrows = true;
+      readVarUnsignedIntTest(testName, bytes, expected, testThrows);
+    }
+
+    // Test all values that can be encoded in a single byte
+    for (let byte = 0; byte <= 0x7F; byte++) {
+        readVarUnsignedInt([byte | 0x80], byte);
+    }
+
+    readVarUnsignedInt([0x80], 0);
+    readVarUnsignedInt([0x00, 0x81], 1);
+    readVarUnsignedInt([0x00, 0x00, 0x81], 1);
+    readVarUnsignedInt([0x00, 0x00, 0x00, 0x81], 1);
+
+    readVarUnsignedInt([0x0E, 0xEB], 1899);
+    readVarUnsignedInt([0x0E, 0xEC], 1900);
```

Comment:
Should we add one for `maxValueForBits(31)`?

Reply:
This would require more complex bit-counting logic.

`maxValueForBits(31)` requires 5 bytes to encode. At 7 usable bits per byte, it's counted as 35 bits even though the 3 most significant bits are 'unused'.